fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java

```diff
@@ -1262,6 +1262,127 @@ private void setJoinPushDown(HashJoinNode node) {
+                    continue;
+                }
+                double cpuCost = rightCost / instanceNum + leftCost / instanceNum;
+                if (cpuCost < cost) {
+                    cost = cpuCost;
+                    bestInstanceNum = instanceNum;
+                    bestDop = 1;
+                    bestDistributionMode = HashJoinNode.DistributionMode.BROADCAST;
+                }
+            }
+
+            // case 3: instance_num=1..defaultDop, dop=defaultDop/instance_num, local shuffle join
+            // CPU_COST: COST(RIGHT)/N/DOP + COST(LEFT)/N/DOP + COST(RIGHT)/N + COST(LEFT)/N
+            // MEM_COST: COST(RIGHT) + COST(LEFT)
+            for (int instanceNum = 1; instanceNum <= defaultDop; ++instanceNum) {
+                for (int dop = defaultDop; dop > 0; --dop) {
+                    memCost = rightCost * instanceNum;
+                    if (memCost > buildBytesSupLimit) {
+                        continue;
+                    }
+                    if (leftCost > probeBytesInfLimit && leftCost / instanceNum > probeBytesInfLimit) {
+                        continue;
+                    }
+                    double cpuCost = rightCost / instanceNum / dop + leftCost / instanceNum / dop
+                            + rightCost / instanceNum + leftCost / instanceNum;
+                    if (cpuCost < cost) {
+                        cost = cpuCost;
+                        bestInstanceNum = instanceNum;
+                        bestDop = dop;
+                        bestDistributionMode = HashJoinNode.DistributionMode.PARTITIONED;
+                    }
+                }
+            }
+
+            // case 4: instance_num=1..defaultDop, dop=defaultDop/instance_num, local broadcast join
+            // CPU_COST: COST(RIGHT)/N + COST(LEFT)/N/DOP
+            // MEM_COST: COST(RIGHT)*N*DOP +COST(LEFT)
+            for (int instanceNum = 1; instanceNum <= defaultDop; ++instanceNum) {
+                for (int dop = defaultDop; dop > 0; --dop) {
+                    memCost = rightCost * instanceNum * dop;
+                    if (memCost > buildBytesSupLimit) {
+                        continue;
+                    }
+                    if (leftCost > probeBytesInfLimit && leftCost / instanceNum > probeBytesInfLimit) {
+                        continue;
+                    }
+                    double cpuCost = rightCost / instanceNum + leftCost / instanceNum / dop;
+                    if (cpuCost < cost) {
+                        cost = cpuCost;
+                        bestInstanceNum = instanceNum;
+                        bestDop = dop;
+                        bestDistributionMode = HashJoinNode.DistributionMode.BROADCAST;
+                    }
+                }
+            }
+            fragment.setParallelExecNum(bestInstanceNum);
+            fragment.setPipelineDop(bestDop);
+            fragment.setDopEstimated();
+            hashJoinNode.setPipelineDistributionMode(bestDistributionMode);
```

Comment:
you couldn't change DistributionMode here.

Reply:
ok